src/Microsoft.AspNet.OData.Shared/Builder/NavigationPropertyExtensions.cs

```diff
@@ -86,19 +95,26 @@ internal static class NavigationPropertyExtensions
                         new Tuple<StructuralTypeConfiguration, IList<MemberInfo>, NavigationPropertyConfiguration>(configuration,
                             bindingPath, nav));
                 }
-                else if (complex != null)
+                else if (complex != null && !typesAlreadyProcessed.Contains(complex.RelatedClrType))
```

Comment:
>!typesAlreadyProcessed.Contains(complex.RelatedClrType)) [](start = 44, length = 56)

Supposed we have "A" type, "B" type is derived from "A".

class A
{
      B b {get;set;}
}

class B : A
{
      A a {get;set;}
}

So, what will happen?

Reply:
I'll add extra test coverage for this case. The short answer is, this scenario works too in the sense that it doesn't infinitely recurse. I'll add test coverage to see what navigation property bindings are produced.

Reply:
Thanks.

---
In reply to: [193264727](https://github.com/OData/webapi/pull/1458#discussion_r193264727) [](ancestors = 193264727)

Reply:
I've added a new test with the above scenario, where `B` has a navigation property `n` and entity type `E` has a property `p` of complex type A. What I observe is that the entity set contains two navigation property bindings:
 1. p/b/n
 1. p/B/n

Note that the first path has only property segments, while the second one has a type-cast segment since property `p` might be of type `B`. This seems like reasonable behavior for this case.